### DISCUSSION TITLE
How to contribute ?

### DISCUSSION THREAD

**Discussion by bevrat** [hklages/node-red-contrib-sonos-plus#252] [Q&A] [answered]
Hello,

I had the need to modify the starting time of an existing Sonos Alarm, without deleting and recreating it.
I've found that the patchAlarm function from sonos API exists and does exactly what I want. I've cloned the repository and modified it on my side. 
I'm totally new to open source contribution (I just used a gitlab repo for myself) and quite rusted with code quality standards. I nevertheless think my use case can interest other people.
How can I help ? 
Should I copy paste the code somewhere and a contributor refactor it, or should I open an issue, and if it's evaluated useful, someone will implement it someday ?

Thanks

**Comment by hklages** [comment, marked as answer]
Hi. 

Ideally it would be:

> clone the repo -> create a new branch -> do your changes -> create a pull request

and then I can directly pull your changes into the project. This usually only works if you use the same IDE, lint settings, style guides, ...  


But as you already had finished your work you can just send me the major part (from sonos-universal.js) and I will check it and most likely publish it in the next release next week.

**Comment by bevrat** [comment]
Ok, so I've succeeded in cloning the repo, create a branch on my side, but I didn't succeed in pushing it to the main repo.
So please find attached the modified files.
[sonos-alarm.zip](https://github.com/hklages/node-red-contrib-sonos-plus/files/10948600/sonos-alarm.zip)

**Comment by hklages** [comment]
well done - looks good. I think I can copy it without any changes. 

Why did you add the helper.js - did you change anything there?

**Reply by bevrat** [reply]
I added a function validTime to check that parameter passed is at a time format, similarly to what was done for validToInteger. I thought it was a convention to separate validations from functions using them.
But it can be moved directly into the householdSetAlarmTime, and simplified, as it's only a regexp check and it's not used by any other function.

**Comment by hklages** [comment]
I copied the code and added a validation of the id and use a different REGEX as the original one do not allowed hours in range from 20 to 23. Have to add two more functions, do the testing and documentation and should finish that in the next 2 - 3 days. 

Thanks for your contribution! Feel free to add more :-)

**Comment by bevrat** [comment]
Thanks, I will now learn how automated testing works. It would have helped me to see the REGEX was not the good one ... 
And I didn't know how to check the existence of alarm id, it's better now :)

**Comment by hklages** [comment]
Automated testing helps a lot. In VSCode I use mocha and chai with npm scripts: `TEST-Helper": "mocha ./test/Helper.js`

> REGEX was not the good one

I changed the REGEX name from `REGEX_TIME` to `REGEX_TIME_SPECIAL`. The name was really misleading. Anyone would expect a hh:mm:ss including 20 .. 24.

**Comment by hklages** [comment]
Hi.
I just publish 6.7.0 - please check whether it works for your and let me know. 

if you would like to create / request any further enhancements, let me also know.

**Reply by bevrat** [reply]
Hi,

Thank you. I've done some testing and it works perfectly.

**Comment by hklages** [comment]
thanks for your feedback and your contribution!
